tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ConcreteFunction.java

```diff
@@ -242,53 +379,245 @@ public void save(String exportDir) throws IOException {
+        .map(ArgDef::getType)
+        .collect(Collectors.toList());
+
+    if (!dataTypesMatch(outputs, nativeOutputs)) {
+      throw new IllegalArgumentException(
+          "Data types of the signature's outputs must match the native function's (in order).  Expected "
+              + nativeOutputs + ", got "
+              + outputs);
+    }
+
+    try (PointerScope scope = new PointerScope()) {
+      this.scope = scope;
+      scope.extend();
+      this.nativeFunction.getNativeHandle().withDeallocatorInScope();
+      this.dependencies.forEach(TF_Function::withDeallocatorInScope);
+    }
   }
 
-  private enum Ownership {
-    GRAPH_AND_SESSION, SESSION_ONLY, NONE;
+  /**
+   * Detects the signature from the handle
+   */
+  static ConcreteFunction fromNativeHandle(NativeFunction nativeFunction,
+      Collection<NativeFunction> availableFunctions) {
+
+    Signature.Builder builder = Signature.builder().methodName(nativeFunction.getFunctionDef().getSignature().getName())
+        .key(nativeFunction.getName());
+
+    for (ArgDef input : nativeFunction.getFunctionDef().getSignature().getInputArgList()) {
+      TensorInfo info = TensorInfo.newBuilder()
+          .setDtype(input.getType())
+          .setTensorShape(TensorShapeProto.newBuilder().setUnknownRank(true).build())
+          .setName(input.getName())
+          .build();
+
+      builder.input(input.getName(), info);
+    }
+
+    for (ArgDef outputDef : nativeFunction.getFunctionDef().getSignature().getOutputArgList()) {
+      TensorInfo info = TensorInfo.newBuilder()
+          .setDtype(outputDef.getType())
+          .setTensorShape(TensorShapeProto.newBuilder().setUnknownRank(true).build())
+          .setName(outputDef.getName())
+          .build();
+
+      builder.output(outputDef.getName(), info);
+    }
+
+    return new ConcreteFunction(
+        builder.build(),
+        nativeFunction,
+        availableFunctions
+    );
   }
 
-  private final Graph graph;
-  private final Session session;
-  private final Signature signature;
-  private final Ownership ownership;
 
-  ConcreteFunction(Signature signature, Graph graph, Session session, Ownership ownership) {
-    this.graph = graph;
-    this.session = session;
-    this.signature = signature;
-    this.ownership = ownership;
+  private static TF_Operation outputHandle(Operand<?> operand) {
+    if (operand == null) {
+      throw new NullPointerException("Can't get output handle for null operand");
+    }
+
+    Pointer handle = operand.asOutput().getUnsafeNativeHandle();
+    if (handle.isNull()) {
+      throw new NullPointerException("Native handle of operand is null, has it been closed?");
+    }
+
+    if (!(handle instanceof TF_Operation)) {
+      throw new IllegalArgumentException("Operand was not a graph operand");
+    }
+
+    return (TF_Operation) handle;
+  }
+
+  private static TF_Output resolveToOutput(Graph graph, List<Operand<?>> operands) {
```

Comment:
Ok I see... so we are now converting all graphs to functions? Is this 100% backward-compatible with the actual solution?

Reply:
It has some issues with initialization since we aren't holding a session, but that's going to happen no matter what as long as we use native functions.  SavedModel loading works fine using the new SavedFunction, and we'll eventually have high level function support + init scopes for creating functions.